core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -814,12 +814,28 @@ private Function compileFilter(
         if (filter != null && !filter.isThreadSafe()) {
             assert filterExpr != null;
             ObjList<Function> workerFilters = new ObjList<>();
-            for (int i = 0; i < workerCount; i++) {
-                restoreWhereClause(filterExpr); // restore original filters in node query models
-                Function workerFilter = compileBooleanFilter(filterExpr, metadata, executionContext);
-                workerFilters.extendAndSet(i, workerFilter);
-                assert filter.getClass() == workerFilter.getClass();
+            boolean supportDeepClone = true;
+            try {
+                for (int i = 0; i < workerCount; i++) {
+                    workerFilters.extendAndSet(i, filter.deepClone());
+                }
+            } catch (UnsupportedOperationException e) {
+                LOG.info().$("Failed to deepClone filter function for parallel worker").I$();
+                supportDeepClone = false;
+            }
+            if (!supportDeepClone) {
+                for (int i = 0, oSize = workerFilters.size(); i < oSize; i++) {
+                    workerFilters.get(i).close();
```

Comment:
Potential NPE here. Should be `Misc.freeObjListAndClear(workerFilters)`. Same with the below changes.